### PR TITLE
sys: Add helpers for pointer tagging

### PR DESCRIPTION
**Pull request by maribu**
#### Background

Pointer tagging refers to (ab)using the unused least significant bits in pointers. E.g. on 32 pointers to word-aligned `struct` (or any other word aligned memory chunk), the two least significant bits have to be zero in any case. Sometimes using these two bits to encode additional data can help saving some RAM.

However, RIOT also supports 8-bit and 16-bit platforms, which have only 0 or 1 bit unused in pointers to word-aligned memory. Without enforcing an alignment to (at least) four bytes, architecture independent pointer tagging is impossible.

#### The Solution

This PR adds an `PTRTAG` attribute that can be used to enforce an 4 byte alignment. This allows using pointer tagging to be used platform independent. And also one helper function for wrapping up a pointer and a two-bit tag into a tagged pointer and two helper function to unwrap the original pointer and the tag for the tagged pointer are also added. This should be a bit more readable than manual bit-fiddling with pointers.

### Testing procedure

The provided test in `tests/sys_ptrtag` should pass. It intentionally uses memory that normally is only byte aligned on all architectures. So if the test passes on a 32 bit arch, it will also pass on 16 bit and 8 bit platforms. (Tested myself on ARM and AVR.)

### Issues/PRs references

None

### CODE REVIEW

**Comment by maribu** [comment]
I was told @fjmolinas is a huge "fan" of nasty dark-magic bit fiddling. Hence, maybe he could have a look into this piece of nasty dark magic?

File: sys/include/ptrtag.h [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+ * @param   ptr         Pointer to tag
+ * @param   tag         Tag to add
+ * @return              Tagged pointer encoding both @p ptr and @p tag
+ *
+ * @pre     @p ptr points to data marked with @ref PTRTAG
+ * @pre     @p tag contains a two bit value (its numeric value 0, 1, 2, or 3)
+ *
+ * Expect assertions blowing up when the preconditions are not met.
+ */
+static inline __attribute__((always_inline)) void * ptrtag(void *ptr, uint8_t tag)
```
[the remark this answers is not in the document]
Comment (maribu) [reply sys/include/ptrtag.h]: Agreed

**Review by benpicco — APPROVED** [review]
This is pretty straightforward.
I don't yet see how this would be used (if you have a struct anyway, might as well extend it by another member), but it's obvious how it works and it's good to have the tool available if they are needed.

Please squash

**Comment by benpicco** [comment]
Oh there is a typo

**Pull request merged**